architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+The application teams are Contoso are allowed to use `Applications.Datastores/redisCaches` and `Applications.Datastores/postgreSQLDatabases`. Suzy uses the `rad` CLI to unregister other types in the `Applications.Datastores` namespace.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+## User Experience (if applicable)
+
+See API design. Tooling and user-experiences are out of scope for this document.
+
+## Design
+
+### High Level Design
+
+The resource type registration APIs allow a user to control the set of resource types known to UCP. Since UCP is responsible for routing, it needs to know the set of resource providers and resource types so that it can send them traffic. 
+
+#### Resource Types
+
+The resource type registration APIs add two new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+`resourceTypes` is our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `resourceTypes` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types. (See scenario 2 above).
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
+
+#### Resource Type Queries
+
+The proposal also defines two new **non-resource** APIs:
+
+- `GET /planes/radius/{plane}/providers/`: list all resource providers.
+- `GET /planes/radius/{plane}/providers/{provider}` shows a specific resource provider.
+
+These APIs are optimized for queries and are consistent with similar APIs in ARM. To understand why this is important, remember that the OpenAPI schema for a resource type can be quite large. Some functionality in Radius will need to query resource types frequently, consider the need for the `rad recipe register` to validate a resource type. Building a separate API for queries allows us to optimize the behavior.
+
+These APIs can also be implemented consistently across all planes in the future. For example Azure doesn't support user-defined types, but it does support looking up the set of resource types.
+
+Lastly, there will still be some system-defined types. For example `System.Resources/resourceProviders` is a system defined type. The query API can return these, but the `System.Resources/resourceProviders` cannot. Imagine for a minute what would happen if users could remove the capability to specify a type.
```

Comment:
I missed understanding this sentence.: For example `System.Resources/resourceProviders` is a system defined type. The query API can return these, but the `System.Resources/resourceProviders` cannot.
I thought we can query system-defined types too.

Reply:
Yes, the query API can return these. The `System.Resources/resourceProviders` API cannot, or at least it can't be modified through `System.Resources/resourceProviders`.